okta/requestExecutor.go

```diff
@@ -73,34 +78,38 @@ func (re *RequestExecutor) NewRequest(method string, url string, body interface{
 }
 
 func (re *RequestExecutor) Do(req *http.Request, v interface{}) (*Response, error) {
-	resp, err := re.httpClient.Do(req)
-
-	if err != nil {
-		return nil, err
+	cacheKey := cache.CreateCacheKey(req)
+	if req.Method != http.MethodGet {
+		re.cache.Delete(cacheKey)
 	}
-	defer resp.Body.Close()
-	response := newResponse(resp)
+	inCache := re.cache.Has(cacheKey)
 
-	err = CheckResponseForError(resp)
-	if err != nil {
-		return response, err
-	}
+	if !inCache {
+		resp, err := re.httpClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
+		defer resp.Body.Close()
 
-	if v != nil {
-		if w, ok := v.(io.Writer); ok {
-			io.Copy(w, resp.Body)
-		} else {
-			decErr := json.NewDecoder(resp.Body).Decode(v)
-			if decErr == io.EOF {
-				decErr = nil
-			}
-			if decErr != nil {
-				err = decErr
-			}
+		respBody, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		origResp := ioutil.NopCloser(bytes.NewBuffer(respBody))
+		resp.Body = origResp
```

Comment:
do consumers of `resp.Body` need to `Seek(0,0)` it before using it?

Also, what are the thread safety / ownership setup of the cache interface with the response objects?

It makes me really wonder if the cache interface should be `[]byte` oriented rather than a full `http.Response`

Reply:
It is the full Response so we have the full object in case we need to do other things with it in the future.  I am afraid that if we only store the splice of bytes of the body, we will have to re-build it in the future if we need some other items from teh response.